src/secret_connection.rs

```diff
@@ -54,7 +56,7 @@ impl<IoHandler: io::Read + io::Write + Send + Sync> SecretConnection<IoHandler>
         let remote_eph_pubkey = share_eph_pubkey(&mut handler, &local_eph_pubkey).unwrap();
 
         // Compute common shared secret.
-        let shared_secret = compute_shared_secret(&remote_eph_pubkey, &local_eph_privkey);
+        let shared_secret = diffie_hellman(&remote_eph_pubkey, &local_eph_privkey);
```

Comment:
It might be worth modeling this in terms of a Noise handshake pattern:

https://latacora.singles/2018/07/18/factoring-the-noise.html

This is a decent start I think, but it'd be nice to e.g. cryptographically bind the signatures that authenticate the connection to the exchange.

Reply:
What tony means is that we could use the signatures as part of the input to the hkdf. This would improve the security properties.

Reply:
So should it go:
```
challenge,recv_key_tmp, send_key_tmp := hkdf(shared_secret)
```
sign challenges encrypted with above tmp keys.
```
recv_key, send_key := hkdf(shared_secret, signature_1, signature_2)
```
where signatures are lexicographically sorted?

Reply:
Why should they be lexographically sorted?

Reply:
I'd need to do some digging into the various options here, but one of them is to sign something that's the equivalent of the transcript hash, i.e. rather than using the signatures as a KDF input, you sign an output.

The Noise equivalent of the TLS transcript hash is the prologue, which is deliberately not mixed into encryption keys.

Reply:
> Why should they be lexographically sorted?

To ensure both parties get the same signature ordering. (You could alternatively use the same ordering provided from the ephemeral pubkeys) Though from Tony's comment, we may do this differently.

I'll read more into noise's key exchange later today. I haven't really looked into it before.